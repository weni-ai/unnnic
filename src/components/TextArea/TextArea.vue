<template>
  <UnnnicFormElement
    :label="label"
    :size="size"
    :disabled="disabled"
    :message="message"
    :error="errors.join(', ') || type === 'error'"
  >
    <textarea
      ref="textarea"
      class="unnnic-text-area__textarea"
      :class="[
        `unnnic-text-area__textarea--size-${size}`,
        `unnnic-text-area__textarea--type-${type}`,
      ]"
      :placeholder="fullySanitize(placeholder)"
      :maxlength="maxLength"
      :disabled="disabled"
      :value="modelValue"
      @input="$emit('update:modelValue', fullySanitize($event.target.value))"
    ></textarea>

    <template
      v-if="maxLength"
      #rightMessage
      >{{ modelValue.length }}/{{ maxLength }}</template
    >
  </UnnnicFormElement>
</template>

<script>
import { fullySanitize } from '../../utils/sanitize';
import UnnnicFormElement from '../FormElement/FormElement.vue';

export default {
  components: {
    UnnnicFormElement,
  },

  props: {
    size: {
      type: String,
      default: 'md',
    },

    label: {
      type: String,
    },

    placeholder: {
      type: String,
    },

    modelValue: {
      type: String,
    },

    maxLength: {
      type: Number,
    },

    message: {
      type: String,
      default: '',
    },

    disabled: {
      type: Boolean,
    },

    type: {
      type: String,
      default: 'normal',
      validator(value) {
        return ['normal', 'error'].indexOf(value) !== -1;
      },
    },

    errors: {
      type: Array,
      default: () => [],
    },
  },
  emits: ['update:modelValue'],

  methods: {
    fullySanitize,
    focus() {
      this.$refs.textarea.focus();
    },
  },
};
</script>

<style lang="scss" scoped>
@use '@/assets/scss/unnnic' as *;
@use '@/components/Input/Input' as *;

.unnnic-text-area {
  &__textarea {
    @include input-base;

    display: block;
    width: 100%;
    resize: vertical;
    box-sizing: border-box;

    scrollbar-width: thin;

    &--size-md {
      @include input-md-font;

      min-height: 6.25 * $unnnic-font-size;
      padding: ($unnnic-spacing-ant - $unnnic-border-width-thinner)
        ($unnnic-spacing-sm - $unnnic-border-width-thinner);
    }

    &--size-sm {
      @include input-sm-font;

      min-height: 5 * $unnnic-font-size;
      padding: ($unnnic-spacing-xs)
        ($unnnic-spacing-sm - $unnnic-border-width-thinner);
    }

    &.unnnic-text-area__textarea--type-error {
      @include input-error;
    }
  }
}
</style>
