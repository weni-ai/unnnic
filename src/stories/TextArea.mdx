import { Meta, Source, Story } from '@storybook/blocks';

import * as TextAreaStories from './TextArea.stories';

<Meta of={TextAreaStories}/>

# TextArea

The `TextArea` component is designed to provide a textarea field for users to enter texts.

<Source
  language='html'
  dark
  code={`<UnnnicTextArea
  label="Label"
  message="Helper text" | error="Error text"
  :disabled="false | true"
  v-model="value"
  size="md"
  type="normal"
/>`}
/>


---

#### **Props Options:**

| Key     | Description                               | Values                  | Default     |
|---------|-------------------------------------------|-------------------------|-------------|
| label   | `String` of the textarea label            | `String`                | `undefined` |
| message | `String` of the helper text, when applied | `String`                | `undefined` |
| error   | `String` of the error text, when applied  | `String`                | `undefined` |
| v-model | `String` of the textarea content          | `''`                    | `''`        |
| size    | `String` of the size                      | `'md'` \| `'sm'`        | `md`        |
| type    | `String` of the textarea type             | `'normal'` \| `'error'` | `'normal'`  |

## Example
Some examples of uses of the `UnnnicTextArea`

### Default

<Story of={TextAreaStories.Default} />


### Error

<Story of={TextAreaStories.Error} />


### Disabled

<Story of={TextAreaStories.Disabled} />


### Default (small size)

<Story of={TextAreaStories.DefaultSm} />


### Error (small size)

<Story of={TextAreaStories.ErrorSm} />


### Disabled (small size)

<Story of={TextAreaStories.DisabledSm} />