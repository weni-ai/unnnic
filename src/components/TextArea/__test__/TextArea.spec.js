import { mount } from '@vue/test-utils';
import { beforeEach, describe, expect, afterEach, test, vi } from 'vitest';
import TextArea from '@/components/TextArea/TextArea.vue';

describe('TextArea.vue', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = mount(TextArea, {
      props: {
        label: 'Description',
        placeholder: 'Enter text',
        modelValue: '',
        maxLength: 200,
        disabled: false,
        type: 'normal',
        size: 'md',
        errors: [],
      },
    });
  });

  afterEach(() => {
    wrapper.unmount();
  });

  test('renders correctly', () => {
    expect(wrapper.exists()).toBe(true);
  });

  test('renders label when provided', () => {
    const label = wrapper.find('.label');
    expect(label.exists()).toBe(true);
    expect(label.text()).toBe('Description');
  });

  test('renders textarea with correct props', () => {
    const textarea = wrapper.find('textarea');
    expect(textarea.attributes('placeholder')).toBe('Enter text');
    expect(textarea.attributes('maxlength')).toBe('200');
    expect(textarea.element.disabled).toBe(false);
  });

  test('renders character count when maxLength and type are normal', () => {
    const helper = wrapper.find('.helper');
    expect(helper.exists()).toBe(true);
    expect(helper.text()).toBe('0/200');
  });

  test('does not render character count when maxLength is not provided', async () => {
    await wrapper.setProps({ maxLength: null });
    const helper = wrapper.find('.helper');
    expect(helper.exists()).toBe(false);
  });

  test('emits update:modelValue on input', async () => {
    const textarea = wrapper.find('textarea');
    await textarea.setValue('new text');
    expect(wrapper.emitted('update:modelValue')[0]).toEqual(['new text']);
  });

  test('displays errors when type is error', async () => {
    await wrapper.setProps({ type: 'error', errors: ['Error 1', 'Error 2'] });
    const errorList = wrapper.findAll('.error-list span');
    expect(errorList.length).toBe(2);
    expect(errorList[0].text()).toBe('Error 1');
    expect(errorList[1].text()).toBe('Error 2');
  });

  test('applies disabled class and disables textarea when disabled is true', async () => {
    await wrapper.setProps({ disabled: true });
    expect(wrapper.classes()).toContain('disabled');
    const textarea = wrapper.find('textarea');
    expect(textarea.element.disabled).toBe(true);
  });

  test('applies the correct size class', async () => {
    await wrapper.setProps({ size: 'sm' });
    expect(wrapper.classes()).toContain('sm');
  });

  test('focus method focuses the textarea element', () => {
    const focusSpy = vi.spyOn(wrapper.vm.$refs.textarea, 'focus');
    wrapper.vm.focus();
    expect(focusSpy).toHaveBeenCalled();
  });

  test('matches the snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  test('does not render character count when type is error', async () => {
    await wrapper.setProps({ type: 'error' });
    const helper = wrapper.find('.helper');
    expect(helper.exists()).toBe(false);
  });

  test('updates character count on input when maxLength is provided', async () => {
    await wrapper.setProps({ modelValue: 'Hello' });
    const helper = wrapper.find('.helper');
    expect(helper.text()).toBe('5/200');
  });

  test('does not render label when not provided', async () => {
    await wrapper.setProps({ label: null });
    const label = wrapper.find('.label');
    expect(label.exists()).toBe(false);
  });

  test('applies correct class when type is "error"', async () => {
    await wrapper.setProps({ type: 'error' });
    expect(wrapper.classes()).toContain('error');
  });

  test('does not emit input when disabled', async () => {
    await wrapper.setProps({ disabled: true });
    const textarea = wrapper.find('textarea');
    await textarea.setValue('new text');
    expect(wrapper.emitted('update:modelValue')).toBeUndefined();
  });

  test('focuses correctly when in error state', async () => {
    await wrapper.setProps({ type: 'error' });
    const focusSpy = vi.spyOn(wrapper.vm.$refs.textarea, 'focus');
    wrapper.vm.focus();
    expect(focusSpy).toHaveBeenCalled();
  });

  test('applies correct class when size is "sm"', async () => {
    await wrapper.setProps({ size: 'sm' });
    expect(wrapper.classes()).toContain('sm');
  });

  test('applies correct class when size is "md"', () => {
    expect(wrapper.classes()).toContain('md');
  });

  test('validates type prop', () => {
    const warnSpy = vi.spyOn(console, 'warn').mockImplementation(() => {});

    mount(TextArea, { props: { type: 'normal' } });
    mount(TextArea, { props: { type: 'error' } });
    expect(warnSpy).not.toHaveBeenCalled();

    mount(TextArea, { props: { type: 'invalid' } });
    expect(warnSpy).toHaveBeenCalled();

    warnSpy.mockRestore();
  });

  test('renders placeholder correctly', () => {
    const textarea = wrapper.find('textarea');
    expect(textarea.attributes('placeholder')).toBe('Enter text');
  });

  test('renders maxLength attribute correctly', () => {
    const textarea = wrapper.find('textarea');
    expect(textarea.attributes('maxlength')).toBe('200');
  });

  test('does not render maxLength attribute when not provided', async () => {
    await wrapper.setProps({ maxLength: undefined });
    const textarea = wrapper.find('textarea');
    expect(textarea.attributes('maxlength')).toBeUndefined();
  });

  test('emits update:modelValue with correct value on input', async () => {
    const textarea = wrapper.find('textarea');
    await textarea.setValue('New value');
    expect(wrapper.emitted('update:modelValue')).toHaveLength(1);
    expect(wrapper.emitted('update:modelValue')[0]).toEqual(['New value']);
  });

  test('does not emit update:modelValue when disabled', async () => {
    await wrapper.setProps({ disabled: true });
    const textarea = wrapper.find('textarea');
    await textarea.setValue('New value');
    expect(wrapper.emitted('update:modelValue')).toBeUndefined();
  });

  test('renders all error messages when type is error', async () => {
    const errors = ['Error 1', 'Error 2', 'Error 3'];
    await wrapper.setProps({ type: 'error', errors });
    const errorMessages = wrapper.findAll('.error-list span');
    expect(errorMessages).toHaveLength(errors.length);
    errorMessages.forEach((msg, index) => {
      expect(msg.text()).toBe(errors[index]);
    });
  });

  test('does not render error list when type is normal', () => {
    const errorList = wrapper.find('.error-list');
    expect(errorList.exists()).toBe(false);
  });

  test('updates character count correctly', async () => {
    await wrapper.setProps({ modelValue: 'Hello, world!' });
    const helper = wrapper.find('.helper');
    expect(helper.text()).toBe('13/200');
  });

  test('focus method focuses the textarea', async () => {
    const textarea = wrapper.find('textarea');
    const focusSpy = vi.spyOn(textarea.element, 'focus');
    wrapper.vm.focus();
    expect(focusSpy).toHaveBeenCalled();
  });
});
