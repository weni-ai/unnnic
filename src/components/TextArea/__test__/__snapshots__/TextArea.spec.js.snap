// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TextArea.vue > matches the snapshot 1`] = `
"<section data-v-9f8d6c86="" data-v-735b3c51="" class="unnnic-form-element">
  <section data-v-7f222291="" data-v-9f8d6c86="" class="unnnic-label unnnic-form-element__label">
    <p data-v-7f222291="" class="unnnic-label__label">Description</p>
    <!--v-if-->
  </section><textarea data-v-735b3c51="" class="unnnic-text-area__textarea unnnic-text-area__textarea--size-md unnnic-text-area__textarea--type-normal" placeholder="Enter text" maxlength="200" value=""></textarea>
  <section data-v-9f8d6c86="" class="unnnic-form-element__hints-container">
    <section data-v-9f8d6c86="" class="unnnic-form-element__message-container">
      <!--v-if-->
      <!--v-if-->
    </section>
    <p data-v-9f8d6c86="">0/200</p>
  </section>
</section>"
`;
